File: src/files.ts
import * as fs from "node:fs";
import * as path from "node:path";
import matter from "gray-matter";
import { safePath } from "./security.js";
import { IKE_DIR, DIRECTORIES } from "./config.js";

// ── Types ─────────────────────────────────────────────────────────────────────

export type TaskStatus = "To Do" | "In Progress" | "Done" | "Draft";
export type Priority = "high" | "medium" | "low";

export interface TaskFrontmatter {
  id: string;
  title: string;
  status: TaskStatus;
  priority?: Priority;
  created: string;
  updated?: string;
  milestone?: string;
  assignees?: string[];
  tags?: string[];
  dependencies?: string[];
  "acceptance-criteria"?: string[];
  "definition-of-done"?: string[];
  visionlog_goal_id?: string;
  blocked_reason?: string;
}

export interface MilestoneFrontmatter {
  id: string;
  title: string;
  status: "open" | "closed";
  created: string;
  due?: string;
}

export interface DocumentFrontmatter {
  id: string;
  title: string;
  created: string;
  updated?: string;
  tags?: string[];
}

export interface ParsedFile<T> {
  frontmatter: T;
  content: string;
  filePath: string;
}

// ── Read / Write ──────────────────────────────────────────────────────────────

export function readMarkdown<T>(filePath: string): ParsedFile<T> {
  const raw = fs.readFileSync(filePath, "utf-8");
  const parsed = matter(raw);
  return { frontmatter: parsed.data as T, content: parsed.content.trim(), filePath };
}

export function writeMarkdown<T extends object>(filePath: string, frontmatter: T, content: string): void {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, matter.stringify(content, frontmatter));
}

// ── ID generation ─────────────────────────────────────────────────────────────

function nextId(prefix: string, existing: string[]): string {
  const max = existing.reduce((acc, id) => {
    const match = id.match(new RegExp(`^${prefix}-(\\d+)$`, "i"));
    if (!match) return acc;
    const n = parseInt(match[1], 10);
    return isNaN(n) ? acc : Math.max(acc, n);
  }, 0);
  return `${prefix}-${String(max + 1).padStart(4, "0")}`;
}

// ── Tasks ─────────────────────────────────────────────────────────────────────

function taskDir(projectRoot: string, completed = false): string {
  return safePath(projectRoot, IKE_DIR, completed ? DIRECTORIES.COMPLETED : DIRECTORIES.TASKS);
}

export function listTasks(projectRoot: string, includeCompleted = false): ParsedFile<TaskFrontmatter>[] {
  const results: ParsedFile<TaskFrontmatter>[] = [];

  const dirs = [taskDir(projectRoot)];
  if (includeCompleted) dirs.push(taskDir(projectRoot, true));

  for (const dir of dirs) {
    if (!fs.existsSync(dir)) continue;
    fs.readdirSync(dir)
      .filter((f) => f.endsWith(".md"))
      .sort()
      .forEach((f) => {
        try {
          results.push(readMarkdown<TaskFrontmatter>(path.join(dir, f)));
        } catch {}
      });
  }
  return results;
}

export function nextTaskId(projectRoot: string): string {
  const existing = listTasks(projectRoot, true).map((t) => t.frontmatter.id);
  // Also check archive
  const archiveDir = safePath(projectRoot, IKE_DIR, DIRECTORIES.ARCHIVE);
  if (fs.existsSync(archiveDir)) {
    fs.readdirSync(archiveDir)
      .filter((f) => f.endsWith(".md"))
      .forEach((f) => {
        try {
          const parsed = readMarkdown<TaskFrontmatter>(path.join(archiveDir, f));
          existing.push(parsed.frontmatter.id);
        } catch {}
      });
  }
  return nextId("TASK", existing);
}

export function taskPath(projectRoot: string, id: string, title: string, completed = false): string {
  const slug = title.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 50);
  return safePath(projectRoot, IKE_DIR, completed ? DIRECTORIES.COMPLETED : DIRECTORIES.TASKS, `${id} - ${slug}.md`);
}

export function findTaskFile(projectRoot: string, id: string): string | null {
  const dirs = [
    safePath(projectRoot, IKE_DIR, DIRECTORIES.TASKS),
    safePath(projectRoot, IKE_DIR, DIRECTORIES.COMPLETED),
    safePath(projectRoot, IKE_DIR, DIRECTORIES.ARCHIVE),
  ];
  for (const dir of dirs) {
    if (!fs.existsSync(dir)) continue;
    const match = fs.readdirSync(dir).find((f) => f.startsWith(id));
    if (match) return path.join(dir, match);
  }
  return null;
}

export function listAllTasks(projectRoot: string): ParsedFile<TaskFrontmatter>[] {
  const results = listTasks(projectRoot, true);
  const archiveDir = safePath(projectRoot, IKE_DIR, DIRECTORIES.ARCHIVE);
  if (fs.existsSync(archiveDir)) {
    fs.readdirSync(archiveDir)
      .filter((f) => f.endsWith(".md"))
      .sort()
      .forEach((f) => {
        try { results.push(readMarkdown<TaskFrontmatter>(path.join(archiveDir, f))); } catch {}
      });
  }
  return results;
}

// ── Milestones ────────────────────────────────────────────────────────────────

function milestoneDir(projectRoot: string): string {
  return safePath(projectRoot, IKE_DIR, DIRECTORIES.MILESTONES);
}

export function listMilestones(projectRoot: string): ParsedFile<MilestoneFrontmatter>[] {
  const dir = milestoneDir(projectRoot);
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter((f) => f.endsWith(".md"))
    .sort()
    .map((f) => readMarkdown<MilestoneFrontmatter>(path.join(dir, f)));
}

export function nextMilestoneId(projectRoot: string): string {
  const existing = listMilestones(projectRoot).map((m) => m.frontmatter.id);
  return nextId("MS", existing);
}

export function milestonePath(projectRoot: string, id: string, title: string): string {
  const slug = title.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 50);
  return safePath(projectRoot, IKE_DIR, DIRECTORIES.MILESTONES, `${id} - ${slug}.md`);
}

export function findMilestoneFile(projectRoot: string, id: string): string | null {
  const dir = milestoneDir(projectRoot);
  if (!fs.existsSync(dir)) return null;
  const match = fs.readdirSync(dir).find((f) => f.startsWith(id));
  return match ? path.join(dir, match) : null;
}

// ── Documents ─────────────────────────────────────────────────────────────────

function documentDir(projectRoot: string): string {
  return safePath(projectRoot, IKE_DIR, DIRECTORIES.DOCUMENTS);
}

export function listDocuments(projectRoot: string): ParsedFile<DocumentFrontmatter>[] {
  const dir = documentDir(projectRoot);
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir)
    .filter((f) => f.endsWith(".md"))
    .sort()
    .map((f) => readMarkdown<DocumentFrontmatter>(path.join(dir, f)));
}

export function nextDocumentId(projectRoot: string): string {
  const existing = listDocuments(projectRoot).map((d) => d.frontmatter.id);
  return nextId("DOC", existing);
}

export function documentPath(projectRoot: string, id: string, title: string): string {
  const slug = title.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 50);
  return safePath(projectRoot, IKE_DIR, DIRECTORIES.DOCUMENTS, `${id} - ${slug}.md`);
}

export function findDocumentFile(projectRoot: string, id: string): string | null {
  const dir = documentDir(projectRoot);
  if (!fs.existsSync(dir)) return null;
  const match = fs.readdirSync(dir).find((f) => f.startsWith(id));
  return match ? path.join(dir, match) : null;
}

File: ike_md/files.py
"""Markdown file I/O, ID generation, task/milestone/document operations."""

import os
import re
from dataclasses import dataclass, field
from typing import Any

import yaml

from .config import IKE_DIR, DIRECTORIES
from .security import safe_path


# ── Types ─────────────────────────────────────────────────────────────────────

@dataclass
class ParsedFile:
    frontmatter: dict[str, Any]
    content: str
    file_path: str


# ── YAML formatting to match gray-matter ──────────────────────────────────────

class _GrayMatterDumper(yaml.SafeDumper):
    """Custom YAML dumper that matches gray-matter (js-yaml) output.

    Key differences from default PyYAML:
    - Strings that look like dates get single-quoted: '2026-03-22'
    - List items are indented 2 spaces under their key (default_flow_style=False
      + best_width handling doesn't do this, we need indent+offset config)
    """
    pass


def _str_representer(dumper: yaml.Dumper, data: str) -> yaml.ScalarNode:
    # Quote strings that look like dates (YYYY-MM-DD)
    if re.match(r"^\d{4}-\d{2}-\d{2}$", data):
        return dumper.represent_scalar("tag:yaml.org,2002:str", data, style="'")
    return dumper.represent_scalar("tag:yaml.org,2002:str", data)


_GrayMatterDumper.add_representer(str, _str_representer)


# ── Read / Write ──────────────────────────────────────────────────────────────

def read_markdown(file_path: str) -> ParsedFile:
    with open(file_path) as f:
        raw = f.read()

    # Parse YAML frontmatter
    if raw.startswith("---\n"):
        end = raw.index("\n---\n", 4)
        fm_str = raw[4:end]
        content = raw[end + 5:].strip()
        frontmatter = yaml.safe_load(fm_str) or {}
        # Convert date objects back to strings (yaml.safe_load parses dates)
        for k, v in frontmatter.items():
            if hasattr(v, "isoformat"):
                frontmatter[k] = v.isoformat()
    else:
        frontmatter = {}
        content = raw.strip()

    return ParsedFile(frontmatter=frontmatter, content=content, file_path=file_path)


def write_markdown(file_path: str, frontmatter: dict[str, Any], content: str) -> None:
    os.makedirs(os.path.dirname(file_path), exist_ok=True)

    # Filter out None values
    fm = {k: v for k, v in frontmatter.items() if v is not None}

    fm_str = yaml.dump(
        fm, Dumper=_GrayMatterDumper,
        default_flow_style=False, sort_keys=False, allow_unicode=True,
        indent=2,  # base indent
    )
    # gray-matter (js-yaml) indents list items under their key with 2 spaces.
    # PyYAML puts them at the same level. Fix by adding 2-space indent to list items
    # that follow a key line.
    lines = fm_str.split("\n")
    fixed = []
    in_list = False
    for line in lines:
        if line.startswith("- "):
            # This is a list item at root level — indent it
            fixed.append("  " + line)
            in_list = True
        else:
            in_list = False
            fixed.append(line)
    fm_str = "\n".join(fixed)

    with open(file_path, "w") as f:
        f.write("---\n")
        f.write(fm_str)
        f.write("---\n")
        if content:
            f.write(content)
            if not content.endswith("\n"):
                f.write("\n")
        else:
            f.write("\n")


# ── ID generation ─────────────────────────────────────────────────────────────

def _next_id(prefix: str, existing: list[str]) -> str:
    max_n = 0
    pattern = re.compile(rf"^{prefix}-(\d+)$", re.IGNORECASE)
    for id_str in existing:
        m = pattern.match(id_str)
        if m:
            n = int(m.group(1))
            max_n = max(max_n, n)
    return f"{prefix}-{str(max_n + 1).zfill(4)}"


def _slugify(title: str, max_len: int = 50) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", title.lower())
    slug = slug.strip("-")
    return slug[:max_len]


# ── Tasks ─────────────────────────────────────────────────────────────────────

def _task_dir(project_root: str, completed: bool = False) -> str:
    subdir = DIRECTORIES["COMPLETED"] if completed else DIRECTORIES["TASKS"]
    return safe_path(project_root, IKE_DIR, subdir)


def list_tasks(project_root: str, include_completed: bool = False) -> list[ParsedFile]:
    results = []
    dirs = [_task_dir(project_root)]
    if include_completed:
        dirs.append(_task_dir(project_root, completed=True))

    for d in dirs:
        if not os.path.exists(d):
            continue
        for f in sorted(os.listdir(d)):
            if f.endswith(".md"):
                try:
                    results.append(read_markdown(os.path.join(d, f)))
                except Exception:
                    pass
    return results


def list_all_tasks(project_root: str) -> list[ParsedFile]:
    results = list_tasks(project_root, include_completed=True)
    archive_dir = safe_path(project_root, IKE_DIR, DIRECTORIES["ARCHIVE"])
    if os.path.exists(archive_dir):
        for f in sorted(os.listdir(archive_dir)):
            if f.endswith(".md"):
                try:
                    results.append(read_markdown(os.path.join(archive_dir, f)))
                except Exception:
                    pass
    return results


def next_task_id(project_root: str) -> str:
    existing = [t.frontmatter["id"] for t in list_tasks(project_root, include_completed=True)]
    # Also check archive
    archive_dir = safe_path(project_root, IKE_DIR, DIRECTORIES["ARCHIVE"])
    if os.path.exists(archive_dir):
        for f in sorted(os.listdir(archive_dir)):
            if f.endswith(".md"):
                try:
                    parsed = read_markdown(os.path.join(archive_dir, f))
                    existing.append(parsed.frontmatter["id"])
                except Exception:
                    pass
    return _next_id("TASK", existing)


def task_path(project_root: str, id: str, title: str, completed: bool = False) -> str:
    slug = _slugify(title)
    subdir = DIRECTORIES["COMPLETED"] if completed else DIRECTORIES["TASKS"]
    return safe_path(project_root, IKE_DIR, subdir, f"{id} - {slug}.md")


def find_task_file(project_root: str, id: str) -> str | None:
    dirs = [
        safe_path(project_root, IKE_DIR, DIRECTORIES["TASKS"]),
        safe_path(project_root, IKE_DIR, DIRECTORIES["COMPLETED"]),
        safe_path(project_root, IKE_DIR, DIRECTORIES["ARCHIVE"]),
    ]
    for d in dirs:
        if not os.path.exists(d):
            continue
        for f in os.listdir(d):
            if f.startswith(id):
                return os.path.join(d, f)
    return None


# ── Milestones ────────────────────────────────────────────────────────────────

def _milestone_dir(project_root: str) -> str:
    return safe_path(project_root, IKE_DIR, DIRECTORIES["MILESTONES"])


def list_milestones(project_root: str) -> list[ParsedFile]:
    d = _milestone_dir(project_root)
    if not os.path.exists(d):
        return []
    return [
        read_markdown(os.path.join(d, f))
        for f in sorted(os.listdir(d))
        if f.endswith(".md")
    ]


def next_milestone_id(project_root: str) -> str:
    existing = [m.frontmatter["id"] for m in list_milestones(project_root)]
    return _next_id("MS", existing)


def milestone_path(project_root: str, id: str, title: str) -> str:
    slug = _slugify(title)
    return safe_path(project_root, IKE_DIR, DIRECTORIES["MILESTONES"], f"{id} - {slug}.md")


def find_milestone_file(project_root: str, id: str) -> str | None:
    d = _milestone_dir(project_root)
    if not os.path.exists(d):
        return None
    for f in os.listdir(d):
        if f.startswith(id):
            return os.path.join(d, f)
    return None


# ── Documents ─────────────────────────────────────────────────────────────────

def _document_dir(project_root: str) -> str:
    return safe_path(project_root, IKE_DIR, DIRECTORIES["DOCUMENTS"])


def list_documents(project_root: str) -> list[ParsedFile]:
    d = _document_dir(project_root)
    if not os.path.exists(d):
        return []
    return [
        read_markdown(os.path.join(d, f))
        for f in sorted(os.listdir(d))
        if f.endswith(".md")
    ]


def next_document_id(project_root: str) -> str:
    existing = [d.frontmatter["id"] for d in list_documents(project_root)]
    return _next_id("DOC", existing)


def document_path(project_root: str, id: str, title: str) -> str:
    slug = _slugify(title)
    return safe_path(project_root, IKE_DIR, DIRECTORIES["DOCUMENTS"], f"{id} - {slug}.md")


def find_document_file(project_root: str, id: str) -> str | None:
    d = _document_dir(project_root)
    if not os.path.exists(d):
        return None
    for f in os.listdir(d):
        if f.startswith(id):
            return os.path.join(d, f)
    return None
